File: spring-data-requery/src/main/java/org/springframework/data/requery/core/RequeryTransactionManager.java
/*
 * Copyright 2018 Coupang Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.data.requery.core;

import io.requery.TransactionIsolation;
import io.requery.sql.EntityDataStore;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.data.annotation.Transient;
import org.springframework.jdbc.datasource.DataSourceTransactionManager;
import org.springframework.transaction.TransactionDefinition;
import org.springframework.transaction.support.DefaultTransactionStatus;

import javax.sql.DataSource;
import java.sql.Connection;

/**
 * Requery용 {@link DataSourceTransactionManager}
 *
 * @author debop
 * @since 18. 6. 14
 * @deprecated Use DataSourceTransactionManager.
 */
@Deprecated
@Slf4j
public class RequeryTransactionManager extends DataSourceTransactionManager {
    private static final long serialVersionUID = 3291422158479490099L;

    @Transient
    private transient EntityDataStore entityDataStore;

    public RequeryTransactionManager(@NotNull final EntityDataStore entityDataStore,
                                     @NotNull final DataSource dataSource) {
        super(dataSource);
        this.entityDataStore = entityDataStore;
    }

    @Override
    protected void doBegin(@NotNull final Object transaction, @NotNull final TransactionDefinition definition) {
        super.doBegin(transaction, definition);
        log.debug("Begin transaction... definition={}", definition);

        TransactionIsolation isolation = getTransactionIsolation(definition.getIsolationLevel());

        if (isolation != null) {
            entityDataStore.transaction().begin(isolation);
        } else {
            entityDataStore.transaction().begin();
        }
    }

    @Override
    protected void doCommit(@NotNull final DefaultTransactionStatus status) {
        if (entityDataStore.transaction().active()) {
            log.debug("Commit transaction. status={}", status.getTransaction());
            entityDataStore.transaction().commit();
        }
        super.doCommit(status);
    }

    @Override
    protected void doRollback(@NotNull final DefaultTransactionStatus status) {
        if (entityDataStore.transaction().active()) {
            log.debug("Rollback transaction. status={}", status);
            entityDataStore.transaction().rollback();
        }
        super.doRollback(status);
    }

    @NotNull
    @Override
    protected Object doSuspend(@NotNull final Object transaction) {
        log.debug("Suspend transaction. transaction={}", transaction);
        return super.doSuspend(transaction);
    }

    @Override
    protected void doResume(@Nullable final Object transaction,
                            @NotNull final Object suspendedResources) {
        log.debug("Resume transaction. transaction={}, suspendedResources={}", transaction, suspendedResources);
        super.doResume(transaction, suspendedResources);
    }

    private @Nullable TransactionIsolation getTransactionIsolation(int isolationLevel) {
        switch (isolationLevel) {
            case Connection.TRANSACTION_NONE:
                return TransactionIsolation.NONE;
            case Connection.TRANSACTION_READ_UNCOMMITTED:
                return TransactionIsolation.READ_UNCOMMITTED;
            case Connection.TRANSACTION_READ_COMMITTED:
                return TransactionIsolation.READ_COMMITTED;
            case Connection.TRANSACTION_REPEATABLE_READ:
                return TransactionIsolation.REPEATABLE_READ;
            case Connection.TRANSACTION_SERIALIZABLE:
                return TransactionIsolation.SERIALIZABLE;

            default:
                return null;
        }
    }
}

File: spring-data-requery-kotlin/src/main/kotlin/org/springframework/data/requery/kotlin/core/RequeryTransactionManager.kt
/*
 * Copyright 2018 Coupang Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.data.requery.kotlin.core

import io.requery.TransactionIsolation
import io.requery.sql.KotlinEntityDataStore
import mu.KotlinLogging
import org.springframework.jdbc.datasource.DataSourceTransactionManager
import org.springframework.transaction.TransactionDefinition
import org.springframework.transaction.support.DefaultTransactionStatus
import java.sql.Connection
import javax.sql.DataSource

/**
 * RequeryTransactionManager
 *
 * @author debop
 * @since 18. 7. 2
 */
// TODO: Spring의 Transaction Isolation 도 반영해야 한다. 
class RequeryTransactionManager(val dataStore: KotlinEntityDataStore<Any>,
                                dataSource: DataSource) : DataSourceTransactionManager(dataSource) {

    private val log = KotlinLogging.logger { }

    override fun doBegin(transaction: Any, definition: TransactionDefinition) {
        super.doBegin(transaction, definition)

        log.debug { "Begin tranaction. definition=$definition" }

        val isolation = transactionIsolationOf(definition.isolationLevel)

        isolation?.let {
            dataStore.transaction.begin(isolation)
        } ?: dataStore.transaction.begin()
    }

    override fun doCommit(status: DefaultTransactionStatus) {
        if(dataStore.transaction.active()) {
            try {
                log.debug { "Commit requery transaction..." }
                dataStore.transaction.commit()
            } catch(ignored: Throwable) {
                log.trace(ignored) { "Fail to commit in requery transaction" }
            }
        }
        super.doCommit(status)
    }

    override fun doRollback(status: DefaultTransactionStatus) {
        if(dataStore.transaction.active()) {
            try {
                log.debug { "Rollback requery transaction..." }
                dataStore.transaction.rollback()
            } catch(ignored: Throwable) {
                log.trace(ignored) { "Fail to rollback in requery transaction" }
            }
        }
        super.doRollback(status)
    }

    override fun doSuspend(transaction: Any): Any {
        log.trace { "Suspend ... transaction=$transaction" }
        return super.doSuspend(transaction)
    }

    override fun doResume(transaction: Any?, suspendedResources: Any) {
        log.trace { "Resume ... transaction=$transaction, suspendedResources=$suspendedResources" }
        super.doResume(transaction, suspendedResources)
    }

    private fun transactionIsolationOf(isolationLevel: Int): TransactionIsolation? = when(isolationLevel) {
        Connection.TRANSACTION_NONE -> TransactionIsolation.NONE
        Connection.TRANSACTION_READ_UNCOMMITTED -> TransactionIsolation.READ_UNCOMMITTED
        Connection.TRANSACTION_READ_COMMITTED -> TransactionIsolation.READ_COMMITTED
        Connection.TRANSACTION_REPEATABLE_READ -> TransactionIsolation.REPEATABLE_READ
        Connection.TRANSACTION_SERIALIZABLE -> TransactionIsolation.SERIALIZABLE
        else -> null
    }
}